pkg/snapshotterutil/sociutil.go

```diff
@@ -0,0 +1,125 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package snapshotterutil
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/containerd/nerdctl/pkg/imgutil"
+	"github.com/sirupsen/logrus"
+)
+
+// CreateSoci creates a SOCI index(`rawRef`)
+func CreateSoci(rawRef string) error {
+	sociExecutable, err := exec.LookPath("soci")
+	if err != nil {
+		logrus.WithError(err).Error("soci executable not found in path $PATH")
+		logrus.Info("you might consider installing soci from: https://github.com/awslabs/soci-snapshotter/blob/main/docs/install.md")
+		return err
+	}
+
+	sociCmd := exec.Command(sociExecutable, []string{"create"}...)
+	sociCmd.Env = os.Environ()
+
+	sociCmd.Args = append(sociCmd.Args, rawRef)
+
+	logrus.Debugf("running %s %v", sociExecutable, sociCmd.Args)
+
+	err = processSociIO(sociCmd)
+	if err != nil {
+		return err
+	}
+
+	return sociCmd.Wait()
+}
+
+// PushSoci pushes a SOCI index(`rawRef`)
+// `hostsDirs` are used to resolve image `rawRef`
+func PushSoci(ctx context.Context, rawRef string, hostsDirs []string) error {
+	digest, err := imgutil.ResolveDigest(ctx, rawRef, false, hostsDirs)
+	if err != nil {
+		logrus.WithError(err).Errorf("unable to resolve digest for an image %s: %v", rawRef, err)
+		return err
+	}
+	ref := rawRef
+	if !strings.Contains(ref, "@") {
```

Comment:
Is there a use-case where this is true? The current call site always contains a specific digest reference, right?

Reply:
You are right. Fixed.